examples/cosmos/ics20/ics20.qnt

```diff
@@ -0,0 +1,308 @@
+
+      // handler.sendPacket
+      val packet = {
+        data: data,
+        sourcePort: sourcePort,
+        sourceChannel: sourceChannel,
+        destPort: "transfer",
+        destChannel: getCounterparty(moduleName, sourceChannel)
+      }
+
+      { moduleState: moduleState.with("bank", bankResult.accounts),
+        inFlightPackets: packets.union(Set(packet)) }
+    } else {
+      { moduleState: moduleState, inFlightPackets: packets }
+    }
+  }
+
+  /// `onRecvPacket` should return an acknowledgment, but it also has to update
+  /// the module state, so the return type is the updated value for the module
+  /// state and the acknowledgment
+  type ReceiveTokenPacketResult = {
+    moduleState: ModuleState,
+    acknowledgement: FungibleTokenPacketAcknowledgement
+  }
+
+  pure def onRecvPacket(moduleState: ModuleState, packet: Packet): ReceiveTokenPacketResult = {
+    val data = packet.data
+    if (movingBackAlongTrace({port: packet.sourcePort, channel: packet.sourceChannel}, data.denom)) {
+      // unescrow tokens to receiver
+      val escrowAccount = moduleState.channelEscrowAddresses.get(packet.destChannel)
+      val receiverDenom = updateTrace(data.denom, packet.sourcePort, packet.sourceChannel, packet.destPort, packet.destChannel)
+
+      val bankResult = moduleState.bank.TransferCoins(escrowAccount, data.receiver, receiverDenom, data.amount)
+
+      val newModuleState = moduleState.with("bank", bankResult.accounts)
+      val ack = if (bankResult.success) {
+        { success: true, errorMessage: "" }
+      } else {
+        { success: false, errorMessage: "transfer coins failed" }
+      }
+
+      { moduleState: newModuleState, acknowledgement: ack }
+    } else {
+      // mint vouchers to receiver
+      val prefix = [packet.destPort, packet.destChannel]
+      val newDenom = updateTrace(data.denom, packet.sourcePort, packet.sourceChannel, packet.destPort, packet.destChannel)
+
+      val bankResult = moduleState.bank.MintCoins(data.receiver, newDenom, data.amount)
+
+      val newModuleState = moduleState.with("bank", bankResult.accounts)
+      val ack = if (bankResult.success) {
+        { success: true, errorMessage: "" }
+      } else {
+        { success: false, errorMessage: "mint coins failed" }
+      }
+
+      { moduleState: newModuleState, acknowledgement: ack }
+    }
+  }
+}
+
+module ics20Test {
+  import base.* from "./base"
+  import ics20.*
+
+  val modules = Set("A", "B", "C")
+
+  /// Set o addresses for random simulation
+  val addresses = Set("alice", "bob", "charlie")
+
+  var moduleStates: str -> ModuleState
+  var inFlightPackets: Set[Packet]
+
+  /// All channels available in a module. Should be extracted to a channels
+  /// quint module.
+  pure def channelsForModule(mod: str): Set[Channel] = {
+    pure val channels = Map(
+      "A" -> Set("channelToB", "channelToC"),
+      "B" -> Set("channelToA", "channelToC"),
+      "C" -> Set("channelToA", "channelToB")
```

Comment:
I don't think the same channel can be used by different modules.
Did you check?

Reply:
I couldn't find anything saying we can't, and the `generateIdentifier` example
given in the [ICS 4
spec](https://github.com/cosmos/ibc/tree/main/spec/core/ics-004-channel-and-packet-semantics#opening-handshake)
makes me think that we could have a `channel-1` in each chain.

If your question was related to the fact that I was calling those "modules"
instead of "chains", than that's fixed.

Reply:
> If your question was related to the fact that I was calling those "modules" instead of "chains", than that's fixed.

That's what threw me off, it's resolved :)